internal/daemon/api_logs.go

```diff
@@ -0,0 +1,318 @@
+
+func v1GetLogs(cmd *Command, _ *http.Request, _ *userState) Response {
+	return logsResponse{
+		svcMgr: cmd.d.overlord.ServiceManager(),
+	}
+}
+
+// Response implementation to serve the logs in a custom format.
+type logsResponse struct {
+	svcMgr serviceManager
+}
+
+func (r logsResponse) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	query := req.URL.Query()
+
+	services := query["services"]
+
+	followStr := query.Get("follow")
+	if followStr != "" && followStr != "true" && followStr != "false" {
+		response := statusBadRequest(`follow parameter must be "true" or "false"`)
+		response.ServeHTTP(w, req)
+		return
+	}
+	follow := followStr == "true"
+
+	numLogs := defaultNumLogs
+	nStr := query.Get("n")
+	if nStr != "" {
+		n, err := strconv.Atoi(nStr)
+		if err != nil {
+			response := statusBadRequest("n must be a valid integer")
+			response.ServeHTTP(w, req)
+			return
+		}
+		numLogs = n
+	}
+
+	// If "services" parameter not specified, fetch logs for all services.
+	if len(services) == 0 {
+		infos, err := r.svcMgr.Services(nil)
+		if err != nil {
+			response := statusInternalError("cannot fetch services: %v", err)
+			response.ServeHTTP(w, req)
+			return
+		}
+		services = make([]string, len(infos))
+		for i, info := range infos {
+			services[i] = info.Name
+		}
+	}
+
+	// Get log iterators by service name (and close them when we're done).
+	itsByName, err := r.svcMgr.ServiceLogs(services, numLogs)
+	if err != nil {
+		response := statusInternalError("cannot fetch log iterators: %v", err)
+		response.ServeHTTP(w, req)
+		return
+	}
+	defer func() {
+		for _, it := range itsByName {
+			_ = it.Close()
+		}
+	}()
+
+	// Output format is JSON Lines, which doesn't have an official mime type,
+	// but "application/x-ndjson" is what most people seem to use:
+	// https://github.com/wardi/jsonlines/issues/9
+	w.Header().Set("Content-Type", "application/x-ndjson")
+
+	// First write n buffered logs
+	if numLogs != 0 {
+		if len(itsByName) == 1 {
+			// Handle single-service case more efficiently
+			for _, it := range itsByName {
+				err = outputLogsSingle(w, it)
+			}
+		} else {
+			err = outputLogsMulti(w, itsByName, numLogs)
+		}
+		if err != nil {
+			fmt.Fprintf(w, "\nerror writing logs: %v", err)
+			return
+		}
+	}
+
+	// If requested, follow and output logs real time until request closed
+	if follow {
+		flushWriter(w)
+		followLogs(w, itsByName, req.Context().Done())
+	}
+}
+
+// Output a single iterator's logs without reading them all into memory.
+func outputLogsSingle(w io.Writer, it servicelog.Iterator) error {
+	reader := bufio.NewReaderSize(it, bufioReaderSize)
+	now := time.Now().UTC()
+	encoder := json.NewEncoder(w)
+	encoder.SetEscapeHTML(false)
+	err := iterateLogs(it, reader, nil, now, func(log logEntry) error {
+		return encoder.Encode(log)
+	})
+	return err
+}
+
+// Output last numLogs logs from multiple iterators (not particularly
+// efficient, as it writes them all to a slice first, then sorts by time).
```

Comment:
When we have N producers generating items that are already ordered inside each of the producers, a more memory conscious implementation and often more efficient as well is to just order the consumption itself.

For example: if we have A, B, and C, we consume one item from each, and keep track of two of them: the earliest one, and the second-to-earliest one. Then, we produce the earliest one and loop in place for as long as that specific producer has timestamps earlier than the second-to-earliest one. Since logs will tend to be clustered and sparse, besides simple and memory efficient, that should also be much faster than doing the entire operation in memory before producing anything.

Another relevant advantage is that with that one single algorithm implementation, we can cover the single case, the multi case, and the follow case, so `outputLogsMulti`, `outputLogsSingle`, and `followLogs` (and maybe `iterateLogs`) can become one single loop.

Without that, the idea of having to allocate and sort dozens or hundreds of thousands of log entries in memory for a trivial operation like this feels disproportionate to the operation being done.

Reply:
Yeah, I think this is fair. It was fairly inefficient. I've reworked the code to do something very close to the above algorithm, merging N streams assuming they're in order (it picks the earliest one each time around). Two things:

1) You can't quite use the same algorithm for the "output last N logs (default)" case and the "output all logs in the ring buffer" case, because we only have a `HeadIterator(N)` which iterates forward. There's currently no way to iterate backward until you have enough (and even if we did have that, you'd still have to buffer them and reverse them). So what I've done is in the "output last N logs" case (the default is 10) I iterate through them as described, and only keep the last 10, using a FIFO queue.
2) For the "follow" case, it's simpler to have a different algorithm, partly because there's a done/cancel channel involved. So I've left `followLogs` largely as it was, though I've changed the outputting to send to a channel to avoid the `Write` concurrency issues you mentioned below.

Reply:
Reassuringly, all of the tests in `api_logs_test.go` still pass after this significant rework.